src/Orleans/IDs/Interner.cs

```diff
@@ -96,31 +96,34 @@ public Interner(int initialSize, TimeSpan cleanupFreq)
         /// <param name="key">key to find</param>
         /// <param name="creatorFunc">function to create new object and store for this key if no cached copy exists</param>
         /// <returns>Object with specified key - either previous cached copy or newly created</returns>
-        public T FindOrCreate(K key, Func<T> creatorFunc)
+        public T FindOrCreate(K key, Func<K, T> creatorFunc)
         {
-            T obj = null;
-            WeakReference cacheEntry = internCache.GetOrAdd(key,
-                (k) =>
-                {
-                    obj = creatorFunc();
-                    return new WeakReference(obj);
-                });
-            if (cacheEntry != null)
+            T result;
+            WeakReference<T> cacheEntry;
+
+            // Attempt to get the existing value from cache.
+            internCache.TryGetValue(key, out cacheEntry);
+
+            // If no cache entry exists, create and insert a new one using the creator function.
+            if (cacheEntry == null)
```

Comment:
@sergeybykov 
> it's not equivalent to the internal implementation of GetOrAdd and can create multiple copies of the cached item

`GetOrAdd` can, too, since it's executed outside of the lock (look at `valueFactory(key)` above). The reason to go with this route over `GetOrAdd` is that this avoids allocations and offers the same semantics/guarantees.